engine/src/main/java/org/terasology/config/flexible/internal/FlexibleConfigImpl.java

```diff
@@ -94,8 +83,18 @@ public boolean remove(SimpleUri id) {
      */
     @Override
     @SuppressWarnings("unchecked")
-    public <V> Setting<V> get(SimpleUri id) {
-        return (Setting<V>) settings.get(id);
+    public <V> Setting<V> get(SimpleUri id, Class<V> valueType) {
+        Setting setting = settings.get(id);
+        Class settingValueClass = setting.getValueClass();
+
+        if (!settingValueClass.equals(valueType)) {
+            throw new ClassCastException(
+                    "Expected a Setting of type " + valueType.getName() +
+                    ", found a Setting of type " + settingValueClass.getName()
+            );
```

Comment:
I just realized that the existing method of retrieving settings from a FC is not type-safe at all; the caller could request a type other than the value type of the Setting being requested and successfully set values without an error. The only way to make it type-safe is to have the `FlexibleConfig.get` method take a `Class<T>` parameter that specifies the type that you want to retrieve the setting as, and then manually check that that type matches the type of the setting looked up.

With that in mind, is this exception message satisfactory?

Reply:
Can we add the name of the setting in there somewhere as well?

Reply:
Sorry, this must have slipped under the radar. I can add the setting id to the error message and push it directly to develop.